src/core-services/account/mutations/createGroup.test.js

```diff
@@ -0,0 +1,98 @@
+import createGroup from "./createGroup.js";
+
+mockContext.getSlug = jest.fn("getSlug");
+mockContext.validatePermissions = jest.fn("validatePermissions");
+mockContext.validatePermissionsLegacy = jest.fn("validatePermissionsLegacy");
+mockContext.collections.Groups.insert = jest.fn("collections.Groups.insert");
+mockContext.collections.Groups.findOne = jest.fn("collections.Groups.findOne");
+mockContext.auth = {
+  upsertRole: jest.fn("auth.upsertRole")
+};
+
+beforeEach (() => jest.resetAllMocks())
+
+test("should create a group for the shop", async () => {
+  const groupId = "test-group-id";
+  const groupName = "test group";
+  const shopId = "test-shop-id";
+  const group = {
+    name: groupName,
+    slug: groupName
+  };
+
+  const input = { group, shopId };
+
+  const fakeResult = {
+    _id: "test-group-id",
+    name: "test group",
+    slug: "test group",
+    permissions: [
+      "dashboard"
+    ],
+    shopId: "test-shop-id"
+  };
+
+  mockContext.collections.Groups.insert.mockReturnValueOnce("test-group-id");
+  mockContext.collections.Groups.findOne
+    .mockReturnValueOnce(fakeResult)
+    .mockReturnValueOnce(undefined)
+    .mockReturnValueOnce(fakeResult);
+
+  mockContext.validatePermissionsLegacy.mockReturnValueOnce(Promise.resolve(undefined));
+  mockContext.validatePermissions.mockReturnValueOnce(Promise.resolve(undefined));
+  mockContext.getSlug.mockReturnValueOnce(groupName);
+  mockContext.auth.upsertRole.mockReturnValueOnce(undefined);
+
+  const result = await createGroup(input, mockContext);
+  const expected = Object.assign({}, { group: fakeResult }, { status: 200 });
+  await expect(result).toEqual(expected);
+
+  expect(mockContext.validatePermissionsLegacy).toHaveBeenCalledWith(["admin"], null, { shopId });
+  expect(mockContext.validatePermissions).toHaveBeenCalledWith("reaction:account", "create", { shopId });
+  expect(mockContext.collections.Groups.findOne).toHaveBeenNthCalledWith(1, { slug: "customer", shopId });
+  expect(mockContext.collections.Groups.findOne).toHaveBeenNthCalledWith(2, { slug: groupName, shopId });
+  expect(mockContext.collections.Groups.findOne).toHaveBeenNthCalledWith(3, { _id: groupId });
+  expect(mockContext.collections.Groups.insert).toHaveBeenCalledWith({
+    name: "test group",
+    slug: "test group",
+    permissions: [
+      "dashboard"
+    ],
+    shopId: "test-shop-id"
+  });
+});
+
+
+test("should throw if group already exists", async () => {
+  const groupName = "test group";
+  const shopId = "test-shop-id";
+  const group = {
+    name: groupName,
+    slug: groupName
+  };
+
+  const input = { group, shopId };
+
+  const fakeResult = {
+    _id: "test-group-id",
+    name: "test group",
+    slug: "test group",
+    permissions: [
+      "dashboard"
+    ],
+    shopId: "test-shop-id"
+  };
+
+  mockContext.collections.Groups.insert.mockReturnValueOnce("test-group-id");
+  mockContext.collections.Groups.findOne
```

Comment:
`insert` and `findOne` should both return a promise. It looks like you have `findOne` returning `undefined` and then later `fakeResult`. If we had to poll for variable changes I don't think any of us here would be fans of JavaScript ;).

Reply:
What I'm saying is the way these tests are written (using `mockReturnValueOnce`) implies the values of functions like `findOne` actually changes. So if you access it once it is `undefined` and if you access again it is `fakeResult`. This means to get the desired `fakeResult` value we would have to poll the variable until it is no longer `undefined`, and that is (very fortunately!) not how promises work.

The result of `findOne` is always a promise. In fact, it is always the _same_ promise. Before the promise resolves it is in a pending state and has no value in its internal slot, and after it resolves it is thrown into a resolved state and is given a value in its slot (which can be `undefined` if that was what it was resolved with).

Jest has great docs on working with promises in tests: https://jestjs.io/docs/en/asynchronous#promises

Also you should look at the other tests in this project to see how we have been working with them. Here is a random file pulled from this codebase: https://github.com/reactioncommerce/reaction/blob/42f24a555e56458b9a7479110430262b92195d41/imports/node-app/core-services/catalog/utils/publishProductToCatalog.test.js#L241-L257

Reply:
@rosshadden ... as discussed, this is meant to return 2 different values on 2 seperate calls i.e. we are not polling as discussed. I have however, changed the return values to be promises as discussed as thats the expected return value from the calls